### PR TITLE
Refactoring dataflow to use emits

### PR DESCRIPTION
**Pull request by nthain**
This is still not fully working. The most recent errors when running remotely are:

Not processing workitem 9086016947210530692 since a deferred exception was found: Traceback (most recent call last): File "/usr/local/lib/python2.7/dist-packages/dataflow_worker/batchworker.py", line 705, in run self._load_main_session(self.local_staging_directory) File "/usr/local/lib/python2.7/dist-packages/dataflow_worker/batchworker.py", line 445, in _load_main_session pickler.load_session(session_file) File "/usr/local/lib/python2.7/dist-packages/apache_beam/internal/pickler.py", line 247, in load_session return dill.load_session(file_path) File "/usr/local/lib/python2.7/dist-packages/dill/dill.py", line 363, in load_session module = unpickler.load() File "/usr/lib/python2.7/pickle.py", line 858, in load dispatch[key](self) File "/usr/lib/python2.7/pickle.py", line 1133, in load_reduce value = func(*args) File "/usr/local/lib/python2.7/dist-packages/dill/dill.py", line 767, in _import_module return getattr(__import__(module, None, None, [obj]), obj) File "/usr/local/lib/python2.7/dist-packages/google/cloud/storage/__init__.py", line 39, in <module> from google.cloud.storage.bucket import Bucket File "/usr/local/lib/python2.7/dist-packages/google/cloud/storage/bucket.py", line 24, in <module> from google.api.core import page_iterator ImportError: No module named core

This does not seem to occur locally for me, so I may need to add a dependency to the setup.py file.

### CODE REVIEW

**Comment by nthain** [comment]
I think I found the missing dependency and added it to setup.py. It seems to be running fine remotely now.

File: ingest_revisions_dataflow/ingest_utils/test_ingester.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+from __future__ import absolute_import
```
Comment (iislucas) [inline ingest_revisions_dataflow/ingest_utils/test_ingester.py]: Add a comment at the top of this file... Is this file supposed to be a test? If so, better to name is ingester_test.py & also to add some actual validity test, e.g. check the the number of output lines is the number expected, and that some of the values are correctly pulled out?

File: ingest_revisions_dataflow/ingest_utils/wikipedia_revisions_ingester.py [hunk 2]
```diff
@@ -33,101 +33,76 @@ def __init__(self, data_reset_path, data_paths, revision_reset_path, out_file):
...
+         self.data['user_text'] = self.data['user_ip']
+      self.data['page_id'] = self.page_data['page_id']
+      self.data['page_namespace'] = self.page_data['page_namespace']
+      self.data['page_title'] = self.page_data['page_title']
+      if not('text' in self.data):
+         self.data['text'] = ""
+      if not('user_id' in self.data):
+         self.data['user_id'] = None
+         self.data['user_text'] = None
+      #if self.data['page_namespace'] in TALK_PAGE_NAMESPACE:
```
Comment (iislucas) [inline ingest_revisions_dataflow/ingest_utils/wikipedia_revisions_ingester.py]: Remove comment?

File: ingest_revisions_dataflow/ingest_utils/run_ingester.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+from __future__ import absolute_import
```
[the remark this answers is not in the document]
Comment (iislucas) [reply ingest_revisions_dataflow/ingest_utils/run_ingester.py]: Also, add a comment at the top of each file to briefly describe what that python file is intended to do.

File: ingest_revisions_dataflow/ingest_utils/run_ingester.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+parser = argparse.ArgumentParser(description='Download and save wikipedia revisions from a dump.')
+parser.add_argument('-i', '--input',  help='Path to a wikipedia xml or 7z revision dump file.')
+
+args = parser.parse_args()
+
+
+def run(input_file = None):
+	
+	data_reset_path, revision_reset_path, data_paths = wiki_ingester.get_paths()
```
Comment (iislucas) [inline ingest_revisions_dataflow/ingest_utils/run_ingester.py]: Why bother exposing these 'internal' looking paths when they are only used again in the  wiki_ingester module? Maybe better to expose just a "parse_stream" function that is given the stream, and does the combination of creating paths and parsing? That or the "ParserContentHandler" gets pulled out into a separate library... ?

File: ingest_revisions_dataflow/dataflow_main.py [hunk 5]
```diff
@@ -13,41 +13,43 @@
...
 
-def run(arg_dict):
+def run(argv = None):
   """Main entry point; defines and runs the wordcount pipeline."""
 
   parser = argparse.ArgumentParser()
   parser.add_argument('--input',
                       dest='input',
-                      default=arg_dict.pop('input'),
+                      default='gs://wikidetox-viz-dataflow/input_lists/7z_file_list_short.txt',
```
[the remark this answers is not in the document]
Comment (nthain) [reply ingest_revisions_dataflow/dataflow_main.py]: There were some issues with my previous implementation of this. I took it out temporarily to fix the larger issues. I'll re-add in a future commit once we have the other parts working.

File: ingest_revisions_dataflow/ingest_utils/run_ingester.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nthain) [reply ingest_revisions_dataflow/ingest_utils/run_ingester.py]: I've added the copyright and license as suggested for all files except setup.py. This I modified from an existing apache beam example that was released under the Apache 2.0 license. I'm not sure how to modify the existing license to fulfill the requirements of the license. In particular, is there a standard convention for the "State changes" clause?

File: ingest_revisions_dataflow/ingest_utils/run_ingester.py (hunk 4 above)
Comment (nthain) [reply ingest_revisions_dataflow/ingest_utils/run_ingester.py]: Good point. Done.

File: ingest_revisions_dataflow/ingest_utils/test_ingester.py (hunk 1 above)
Comment (nthain) [reply ingest_revisions_dataflow/ingest_utils/test_ingester.py]: Done.

File: ingest_revisions_dataflow/ingest_utils/wikipedia_revisions_ingester.py (hunk 2 above)
Comment (nthain) [reply ingest_revisions_dataflow/ingest_utils/wikipedia_revisions_ingester.py]: Done. Now it only writes talk pages which should save file size enormously.

**Comment by nthain** [comment]
Hey Lucas, I addressed your comments above. If this looks good, give me an LGTM and I'll merge.

File: ingest_revisions_dataflow/ingest_utils/ingester_test.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import unittest
+import json
+import xml.sax
+import subprocess
+from os import path
+from ingest_utils import wikipedia_revisions_ingester as wiki_ingester
+
+class TestWikiIngester(unittest.TestCase):
+
+	def test_ingester(self):
```
Comment (iislucas) [inline ingest_revisions_dataflow/ingest_utils/ingester_test.py]: Use 2-space indent.

Comment (nthain) [reply ingest_revisions_dataflow/ingest_utils/ingester_test.py]: Done.

**Pull request merged**